### ISSUE TITLE
use LLM for alternatives

### ISSUE THREAD

**Issue by lsmith77** [witty-works/nlp_api#1086]
Algorithm

-
    - return the original alternative lemma form to the browser

- return the sentence for the highlighted phrase from the \/check API \(with start\/end\)

- when user opens popup, we fetch LLM based alternatives via a new \/rephrase API

- on hover we show the grammatically correct alternative in the rewritten sentence \(potentially highlighting what was changed\)

![Image](https://github.com/witty-works/nlp_api/assets/300279/19a25b8f-c67a-49b4-80c7-c7ec1917d964)

Note: make sure genderstar alternatives are not scewed up \(f\.e "Hallo Kunde" or "Hallo Vorgesetzter"\)

<https://www.notion.so/witty-works/LLMs-d1897e284afe4dd98c8ca8555d8c57c8#1455ccd8f74f48d5a79d7b363a076620>

<p></p>

Open questions:

- How to determine the sentence?

    - Server side inside \/check?

    - Client side?

- How to handle remove alternatives?

**Comment by lsmith77** [comment]
Link to feature: https://wittyworks.productboard.com/feature-board/planning/features/27265908

**Comment by lsmith77** [comment]
* ~~EU hosting is a must (@lsmith77 )~~ (see below)
    * Find list of self-host able ? Use as a tiebreaker
* Project how many credits will be used (@lsmith77 )
    * Is there a difference is credits used?
        * Yes https://aws.amazon.com/bedrock/pricing/
* ~~Bedrock setup~~ (@lsmith77 )
    * Boto3
        * Figure out how to create credential file
* Test the prompt on different models (@hellasol )
    * Test
        * English
        * German
        * French

**Comment by lsmith77** [comment]
# AI21 Labs - Jurassic-2

EU Hosting: ?
Self-hostable: No

# Amazon - Titan

EU Hosting: Yes (Frankfurt, Ireland, London(?), Paris)
Self-hostable: No 

# Anthropic - Claude 2/3

EU Hosting: Yes (Frankfurt, Paris)
Self-hostable: No (https://news.ycombinator.com/item?id=36138224)

# Meta - Llama 2/3

**EU Hosting: Yes (Ireland)**
**Self-hostable: Yes, https://llama.meta.com/docs/llama-everywhere**

# Mistral AI - Mistral

**EU Hosting: Yes (Ireland, Paris)**
**Self-hostable: Yes, https://docs.mistral.ai/deployment/self-deployment/overview/**

**Comment by lsmith77** [comment]
https://medium.com/@neylsoncrepalde/how-to-use-amazon-bedrock-to-write-an-article-about-amazon-bedrock-ff09329e9734

https://boto3.amazonaws.com/v1/documentation/api/latest/guide/credentials.html

**Comment by hellasol** [comment]
I ended up running it in the playground as each model has its own params etc.. It also has a handy model comparison feature. 

I focused on llama and mistral models. 

I found that for english we could get away with:



Llama 3 70B Instruct 
Llama 2 70 B 
Mistral 7B Instruct 
Mixtral 8x7B Instruct 
Mistral Large 

For german: 
Llama 3 70B Instruct 
Mixtral 8x7B Instruct  
Mistral Large

Overall Mistral seems more stable to me and works much better than llama with smaller models (huge difference Mistral 7B Instruct 
and Llama 3 8B Instruct).
For german and more complicated senetences the bigger models were obviously better.
Mixtral 8x7B performed surprisingly well and its much cheaper than the larger ones. So maybe its worth testing this one well with different prompts to see it its good enough? 

![Image](https://github.com/witty-works/nlp_api/assets/45122311/cb5ff920-3f16-44a0-8de5-842fb3295941)
![Image](https://github.com/witty-works/nlp_api/assets/45122311/9a3ff2da-ce3f-4dd0-b6a7-45094832f929)
